### PR TITLE
JABERA-10 add likes for posts

### PR DESCRIPTION
**Pull request by jkk84**

### CODE REVIEW

File: src/main/java/com/the/dark/side/crew/fejsbuk/controller/LikeController.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/likes")
+public class LikeController {
+
+    private final LikeService likeService;
+
+    @PostMapping
+    public ResponseEntity<String> addLike
```
Comment (rmaj91) [inline src/main/java/com/the/dark/side/crew/fejsbuk/controller/LikeController.java]: formatting

File: src/main/java/com/the/dark/side/crew/fejsbuk/service/impl/LikeServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        UserEntity userEntity = userRepository.findById(userId)
+                .orElseThrow(() -> new ResponseStatusException
+                        (HttpStatus.NOT_FOUND, "User " + userId + " not found."));
+        postEntity.getLikes().add(userEntity);
+        postRepository.save(postEntity);
+        return "User " + userId + " like post " + postId + ".";
+    }
+
+    @Override
+    public Map<String, Integer> countLikes(long postId) {
```
Comment (rmaj91) [inline src/main/java/com/the/dark/side/crew/fejsbuk/service/impl/LikeServiceImpl.java]: why map?

File: src/main/java/com/the/dark/side/crew/fejsbuk/service/impl/LikeServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    public String addLike(long postId, long userId) {
+        PostEntity postEntity = postRepository.findById(postId)
+                .orElseThrow(() -> new ResponseStatusException
+                        (HttpStatus.NOT_FOUND, "Post " + postId + " not found."));
+        UserEntity userEntity = userRepository.findById(userId)
+                .orElseThrow(() -> new ResponseStatusException
+                        (HttpStatus.NOT_FOUND, "User " + userId + " not found."));
+        postEntity.getLikes().add(userEntity);
+        postRepository.save(postEntity);
+        return "User " + userId + " like post " + postId + ".";
```
Comment (rmaj91) [inline src/main/java/com/the/dark/side/crew/fejsbuk/service/impl/LikeServiceImpl.java]: why?

File: src/main/java/com/the/dark/side/crew/fejsbuk/service/impl/LikeServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+import org.springframework.stereotype.Service;
+import org.springframework.web.server.ResponseStatusException;
+
+import java.util.Collections;
+import java.util.Map;
+
+
+@Service
+@RequiredArgsConstructor
+public class LikeServiceImpl implements LikeService {
```
Comment (rmaj91) [inline src/main/java/com/the/dark/side/crew/fejsbuk/service/impl/LikeServiceImpl.java:16]: missing tests

File: src/main/java/com/the/dark/side/crew/fejsbuk/model/PostEntity.java [hunk 5]
```diff
@@ -25,5 +26,11 @@
 
     @ManyToOne
     private UserEntity userEntity;
-}
 
+    @ManyToMany
```
Comment (rmaj91) [inline src/main/java/com/the/dark/side/crew/fejsbuk/model/PostEntity.java]: Why ManyToMany?

**Review by rmaj91 — COMMENTED** [review]
.

File: src/main/java/com/the/dark/side/crew/fejsbuk/controller/PostController.java [hunk 6]
```diff
@@ -12,6 +12,7 @@
 @RestController
 @RequiredArgsConstructor
 @RequestMapping("/posts")
+@CrossOrigin("*")
```
Comment (rmaj91) [inline src/main/java/com/the/dark/side/crew/fejsbuk/controller/PostController.java]: remove this

File: src/main/java/com/the/dark/side/crew/fejsbuk/service/impl/LikeServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+    private final LikeRepository likeRepository;
+    private final LikeMapper likeMapper;
+
+    @Override
+    public LikeDto addLike(LikeDto likeDto) {
+        LikeEntity likeEntity = likeMapper.toEntity(likeDto);
+        if(likeRepository.existsByUserEntityAndPostEntity(likeEntity.getUserEntity(), likeEntity.getPostEntity())) {
+            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Already liked.");
```
Comment (rmaj91) [inline src/main/java/com/the/dark/side/crew/fejsbuk/service/impl/LikeServiceImpl.java]: i will suggest CONFLICT

File: src/main/java/com/the/dark/side/crew/fejsbuk/service/impl/LikeServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+@Service
+@RequiredArgsConstructor
+public class LikeServiceImpl implements LikeService {
+
+    private final LikeRepository likeRepository;
+    private final LikeMapper likeMapper;
+
+    @Override
+    public LikeDto addLike(LikeDto likeDto) {
+        LikeEntity likeEntity = likeMapper.toEntity(likeDto);
```
Comment (rmaj91) [inline src/main/java/com/the/dark/side/crew/fejsbuk/service/impl/LikeServiceImpl.java:23]: mapping after validation, user id's for valiation

File: src/main/java/com/the/dark/side/crew/fejsbuk/service/impl/LikeServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Already liked.");
+        } else likeRepository.save(likeEntity);
+        return likeMapper.toDto(likeEntity);
+    }
+
+    @Override
+    public void removeLike(long likeId) {
+        if(likeRepository.existsById(likeId)) {
+            likeRepository.deleteById(likeId);
+        } else throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Like " + likeId + " does not exist.");
```
Comment (rmaj91) [inline src/main/java/com/the/dark/side/crew/fejsbuk/service/impl/LikeServiceImpl.java]: use { }

File: src/main/java/com/the/dark/side/crew/fejsbuk/service/impl/LikeServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+    private final LikeRepository likeRepository;
+    private final LikeMapper likeMapper;
+
+    @Override
+    public LikeDto addLike(LikeDto likeDto) {
+        LikeEntity likeEntity = likeMapper.toEntity(likeDto);
+        if(likeRepository.existsByUserEntityAndPostEntity(likeEntity.getUserEntity(), likeEntity.getPostEntity())) {
+            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Already liked.");
+        } else likeRepository.save(likeEntity);
```
Comment (rmaj91) [inline src/main/java/com/the/dark/side/crew/fejsbuk/service/impl/LikeServiceImpl.java]: {}

File: src/main/java/com/the/dark/side/crew/fejsbuk/service/impl/LikeServiceImpl.java (hunk 9 above)
Comment (rmaj91) [reply src/main/java/com/the/dark/side/crew/fejsbuk/service/impl/LikeServiceImpl.java]: https://stackoverflow.com/questions/12193170/whats-the-purpose-of-using-braces-i-e-for-a-single-line-if-or-loop

File: src/test/java/com/the/dark/side/crew/fejsbuk/mapper/LikeMapperTest.java [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+    }
+
+    @Test
+    void whenToDtoThenReturnLikeDto() {
+        LikeEntity likeEntity = mock(LikeEntity.class);
+        PostEntity postEntity = mock(PostEntity.class);
+        UserEntity userEntity = mock(UserEntity.class);
+        LocalDateTime createdAt = LocalDateTime.now();
+        long likeId = 3L;
+        long postId = 1L;
```
Comment (rmaj91) [inline src/test/java/com/the/dark/side/crew/fejsbuk/mapper/LikeMapperTest.java]: you can extract this variables to static class variable and use it in all tests

File: src/test/java/com/the/dark/side/crew/fejsbuk/mapper/LikeMapperTest.java [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+
+        ResponseStatusException exception =
+                assertThrows(ResponseStatusException.class, () -> likeMapper.toEntity(likeDto));
+        assertEquals("404 NOT_FOUND \"User 2 not found.\"", exception.getMessage());
+    }
+
+    @Test
+    void whenPostNotFoundThenReturn404() {
+        long postId = 2L;
+        long userId = 2L;
```
Comment (rmaj91) [inline src/test/java/com/the/dark/side/crew/fejsbuk/mapper/LikeMapperTest.java]: bad test, If you would switch userId and postId in tested method test will still pass.

File: src/test/java/com/the/dark/side/crew/fejsbuk/service/impl/LikeServiceImplTest.java [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+
+    @Test
+    void shouldRemoveLike() {
+        long likeId = 1L;
+        LikeEntity likeEntity = mock(LikeEntity.class);
+
+        when(likeEntity.getId()).thenReturn(likeId);
+
+        likeRepository.deleteById(likeEntity.getId());
+
```
Comment (rmaj91) [inline src/test/java/com/the/dark/side/crew/fejsbuk/service/impl/LikeServiceImplTest.java]: you can delete this line

File: src/test/java/com/the/dark/side/crew/fejsbuk/service/impl/LikeServiceImplTest.java [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+        ResponseStatusException exception =
+                assertThrows(ResponseStatusException.class, () -> likeService.removeLike(likeId));
+        assertEquals("404 NOT_FOUND \"Like 1 does not exist.\"", exception.getMessage());
+    }
+
+    @Test
+    void countLikes() {
+        long postId = 1L;
+        long count = 1L;
```
Comment (rmaj91) [inline src/test/java/com/the/dark/side/crew/fejsbuk/service/impl/LikeServiceImplTest.java:100]: identical variables

Comment (rmaj91) [reply src/test/java/com/the/dark/side/crew/fejsbuk/service/impl/LikeServiceImplTest.java:100]: values*

**Review by rmaj91 — COMMENTED** [review]
Well Done! Almost!

File: src/test/java/com/the/dark/side/crew/fejsbuk/service/impl/LikeServiceImplTest.java (hunk 13 above)
Comment (jkk84) [reply src/test/java/com/the/dark/side/crew/fejsbuk/service/impl/LikeServiceImplTest.java]: without this line test failed, any idea?

Comment (rmaj91) [reply src/test/java/com/the/dark/side/crew/fejsbuk/service/impl/LikeServiceImplTest.java]: line nr 74? :D

File: src/main/java/com/the/dark/side/crew/fejsbuk/controller/LikeController.java [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+
+    @PostMapping
+    public ResponseEntity<LikeDto> addLike(@RequestBody LikeDto likeDto) {
+        return ResponseEntity.ok(likeService.addLike(likeDto));
+    }
+
+    @DeleteMapping
+    public ResponseEntity<Void> removeLike(@RequestParam long likeId) {
+        likeService.removeLike(likeId);
+        return new ResponseEntity<>(HttpStatus.OK);
```
Comment (rmaj91) [inline src/main/java/com/the/dark/side/crew/fejsbuk/controller/LikeController.java]: You dont have to send ok. This is default status

**Review by rmaj91 — APPROVED** [review]
Can merge, later we will adapt this chages to jwt auth, while frontent work

**Pull request merged**